src/main/java/uk/gov/justice/hmpps/prison/service/receiveandtransfer/PrisonerCreationService.kt

```diff
@@ -0,0 +1,191 @@
+  private val offenderIdentifierRepository: OffenderIdentifierRepository,
+  private val offenderRepository: OffenderRepository,
+  private val offenderTransformer: OffenderTransformer,
+  private val prisonerRepository: PrisonerRepository,
+) {
+  fun createPrisoner(requestToCreate: RequestToCreate): InmateDetail {
+    val gender: Gender = gender(requestToCreate.gender).getOrThrow()
+    val ethnicity: Ethnicity? = ethnicity(requestToCreate.ethnicity)?.getOrThrow()
+    val title: Title? = title(requestToCreate.title)?.getOrThrow()
+    val suffix: Suffix? = suffix(requestToCreate.suffix)?.getOrThrow()
+
+    val validPncNumber: String? = validPncNumber(requestToCreate.pncNumber)?.getOrThrow()
+    val validCroNumber: String? = validCroNumber(requestToCreate.croNumber)?.getOrThrow()
+    val (firstName, lastName) = validUniqueUppercaseNames(requestToCreate).getOrThrow()
+    val dateOfBirth = validDateOfBirth(requestToCreate.dateOfBirth).getOrThrow()
+
+    val prisoner = offenderRepository.save(
+      Offender()
+        .withLastName(lastName)
+        .withFirstName(firstName)
+        .withMiddleName(requestToCreate.middleName1?.uppercase())
+        .withMiddleName2(requestToCreate.middleName2?.uppercase())
+        .withBirthDate(dateOfBirth)
+        .withGender(gender)
+        .withTitle(title)
+        .withSuffix(suffix)
+        .withEthnicity(ethnicity)
+        .withCreateDate(LocalDate.now())
+        .withNomsId(getNextPrisonerIdentifier().getId())
+        .withIdSourceCode("SEQ")
+        .withNameSequence("1234")
+        .withCaseloadType("INST")
+        .withLastNameKey(lastName)
+        .withLastNameAlphaKey(lastName.take(1))
+        .withLastNameSoundex(Soundex().soundex(lastName))
+    ).also { newPrisoner ->
+      newPrisoner.rootOffenderId = newPrisoner.id
+      validPncNumber?.let { newPrisoner.addIdentifier("PNC", it) }
+      validCroNumber?.let { newPrisoner.addIdentifier("CRO", it) }
+    }
+
+    return offenderTransformer.transform(prisoner)
+  }
+
+  fun getNextPrisonerIdentifier(): PrisonerIdentifier {
+    var retries = 0
+    var updated: Boolean
+    var nextSequence: NomsIdSequence
+    var currentSequence: NomsIdSequence
+    do {
+      currentSequence = prisonerRepository.nomsIdSequence
+      nextSequence = currentSequence.next()
+      updated = prisonerRepository.updateNomsIdSequence(nextSequence, currentSequence) > 0
+    } while (!updated && retries++ < 10)
+    if (!updated) {
+      throw RuntimeException("Prisoner Identifier cannot be generated, please try again")
+    }
+    return PrisonerIdentifier().withId(currentSequence.prisonerIdentifier)
+  }
+
+  private fun gender(code: String): Result<Gender> =
+    genderRepository.findByIdOrNull(ReferenceCode.Pk(Gender.SEX, code))?.let { success(it) } ?: failure(
+      EntityNotFoundException.withMessage("Gender $code not found")
+    )
+
+  private fun ethnicity(code: String?): Result<Ethnicity>? =
```

Comment:
I'm not sure why these are nullable? Could return type to be `Result<Ethnicity?>`?